File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    diff = []
    
    def dfs(node, graph, visited):
        visited[node] = True
        count = 1
        for n in graph[node]:
            if not visited[n]:
                count += dfs(n,graph, visited)
        return count
    
    #인접리스트로 표시
    graph = [[] for _ in range(n+1)]
    for a,b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    #전선 하나 끊기
    for i in range(len(wires)):
        temp_graph = [row[:] for row in graph] #깊은복사
        
        a,b = wires[i] #끊을 전선
        
        #전선 끊기
        temp_graph[a].remove(b)
        temp_graph[b].remove(a)
        
        #한쪽 노드 개수 세기
        visited = [False]*(n+1)
        tree1 = dfs(a, temp_graph, visited)
        tree2 = n- tree1
        diff.append(abs(tree1-tree2))
        
    return min(diff)
        

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<vector<int>> wires) {
    // 최댓값으로 초기화
    int answer = n;
    
    // 하나씩 wire 끊어서 인접리스트 만들어서 표현
    for(int no = 0 ; no < wires.size(); no++){
        
        // 1. 인접리스트 만들기
        vector<vector<int>> graph(n+1);
        
        for (vector<int> wire : wires){ //auto로 분해는 tuple, pair만 가능
            int a = wire[0];
            int b = wire[1];
            
            // 끊은 wire에 대해서는 그래프를 생성하지 않음
            if ( (a == wires[no][0] && b == wires[no][1])){
                continue;
            }
            graph[a].push_back(b); //이렇게 접근하려면 graph 초기화했어야 함 
            graph[b].push_back(a);
        }
        
        // 2. graph 돌면서 연결된 k 개의 노드 찾기
        queue<int> q;
        q.push(1);
        int k = 1;
        
        vector<bool> visited(n+1, false);
        visited[1]= true;
        
        while(q.size() > 0){
            int cur = q.front();
            q.pop();
            
            for(int node: graph[cur]){
                if(!visited[node]){
                    q.push(node);
                    visited[node] = true;
                    k++; //연결된 노드 하나 추가
                }
                
            }
        }
        // 3. 차이값이 최소이면 answer 갱신
        answer = min(abs(k - (n-k)), answer);
    }
    return answer;
}